tests/scripts/run-xunit-perf.py

```diff
@@ -0,0 +1,466 @@
+parser.add_argument('-library', dest='library', action='store_true', default=False)
+parser.add_argument('-generateBenchviewData', dest='benchviewPath', default=None)
+parser.add_argument('-arch', dest='arch', default='x64')
+parser.add_argument('-os', dest='operatingSystem', default=sys.platform)
+parser.add_argument('-optLevel', dest='optLevel', default='full_opt')
+parser.add_argument('-jitName', dest='jitName', default='ryujit')
+parser.add_argument('-noPgo', dest='isPgoOptimized', action='store_false', default=True)
+parser.add_argument('-configuration', dest='configuration', default='Release')
+parser.add_argument('-group', dest='benchviewGroup', default='CoreCLR')
+parser.add_argument('-outputdir', dest='outputDir', default=None)
+
+##########################################################################
+# Helper Functions
+##########################################################################
+
+def validate_args(args):
+    """ Validate all of the arguments parsed.
+    Args:
+        args (argparser.ArgumentParser): Args parsed by the argument parser.
+    Returns:
+        (arch, build_type, clr_root, fx_root, fx_branch, fx_commit, env_script)
+            (str, str, str, str, str, str, str)
+    Notes:
+    If the arguments are valid then return them all in a tuple. If not, raise
+    an exception stating x argument is incorrect.
+    """
+
+    def validate_arg(arg, check):
+        """ Validate an individual arg
+        Args:
+           arg (str|bool): argument to be validated
+           check (lambda: x-> bool): test that returns either True or False
+                                   : based on whether the check passes.
+
+        Returns:
+           is_valid (bool): Is the argument valid?
+        """
+
+        helper = lambda item: item is not None and check(item)
+
+        if not helper(arg):
+            raise ValueError('Argument: %s is not valid.' % (arg))
+
+    valid_arches = ['x64', 'x86']
+    valid_jits = ['ryujit']
+    valid_optlevels = ['full_opt', 'min_opt', 'tiered']
+    valid_better = ['desc', 'asc']
+    valid_runtypes = ['rolling', 'private', 'local']
+    valid_configurations = ['Release', 'Checked', 'Debug']
+    valid_operatingSystems = ['Windows_NT', 'Ubuntu16.04', 'Ubuntu14.04', 'OSX', sys.platform]
+
+    coreclrPerf = os.path.join(os.getcwd(), args.coreclrPerf)
+
+    validate_arg(args.arch, lambda item: item in valid_arches)
+    validate_arg(args.operatingSystem, lambda item: item in valid_operatingSystems)
+    validate_arg(args.jitName, lambda item: item in valid_jits)
+    validate_arg(args.optLevel, lambda item: item in valid_optlevels)
+    validate_arg(coreclrPerf, lambda item: os.path.isdir(item))
+    validate_arg(args.better, lambda item: item in valid_better)
+    validate_arg(args.runType, lambda item: item in valid_runtypes)
+    validate_arg(args.configuration, lambda item: item in valid_configurations)
```

Comment:
`argparse` has an option for options that have a fixed set of valid values.
```python
parser.add_argument('--option',help='Help string for this option.', required=True, choices=['choice-1', 'choice-2'])
```

Reply:
I did not realize that. I will update this.

Reply:
`validate_arg` can be replaced with the `type=` argument to the `add_argument` function.
```python
def type_arg_validator(arg):
    # handle it

parser.add_argument('--option', help='Help description.', type=type_arg_validator, ... )
```
I'm not sure if lambda works there :)

---
In reply to: [160546138](https://github.com/dotnet/coreclr/pull/15568#discussion_r160546138) [](ancestors = 160546138)

Reply:
Honestly, this is how we've done it in other scripts, hence how I did it here. If it's something you feel strongly about, I'll look into it, but otherwise, I'd rather just leave it as is.